### ISSUE TITLE
Feature Request: Filter/Toggle images in search

### ISSUE THREAD

**Issue by Petri-Oosthuizen** [p0deje/Maccy#206]
Add the ability to show only images as you can't currently easily find images.

**Comment by p0deje** [comment]
Not sure how exactly this should look like and behave.
I don't really have time to invest in this now, but PRs are always welcome!

**Comment by kidonng** [comment]
I think [Alfred](https://www.alfredapp.com/help/features/clipboard/)'s approach is quite intuitive: they show image entries as `Image: <width>x<height> (<size>)` in the clipboard item list (for files they use a similar format `File: <filename>`). Therefore, they can be searched with `Image` or `File`, just like text items.

**Comment by p0deje** [comment]
I am trying to implement a way to filter based on the type and I can't seem to find a good UI/UX for it. I'd like to share a few examples and hopefully, you can react with 👍 / 👎 if you like or don't like the solution.

In this example, I use segmented control where segments show or hide Files / Images / Text.

<img width="440" alt="Screen Shot 2022-08-13 at 12 48 14" src="https://user-images.githubusercontent.com/665846/184508662-80d5bb34-25ec-4c28-b6cd-6cd6b8b5e265.png">

**Comment by p0deje** [comment]
Another possible solution would be to have one button which shows a popover on click. This popover would have 3 checkboxes for Files / Images / Text.

<img width="642" alt="Screen Shot 2022-08-13 at 13 29 06" src="https://user-images.githubusercontent.com/665846/184509802-016f4f94-41d3-4fdf-ab90-be8502e49499.png">

**Comment by Petri-Oosthuizen** [comment]
I like the first option, it clearly shows the current filter state. Unfortunately, it makes the search bar header a bit busier - I’m ok with that though. If it becomes too busy the toggles could always be moved to underneath the search bar.

The second option is a cleaner approach but requires more clicks to toggle states and state is not obvious.

**Comment by p0deje** [comment]
I lean towards implementing something like this. In this approach it would:
1. By default show all types of data.
2. Click on a single type would hide/show this data.
3. Click on a single type with pressed Option would show ONLY this data.
4. It should be possible to switch to the control by pressing "Tab" and then using the left/right arrows to select type and Space (Option-Space) to simulate clicking on the button.

<img width="489" alt="Screen Shot 2022-08-22 at 06 32 27" src="https://user-images.githubusercontent.com/665846/185933457-dc159de4-9d25-4053-8558-c7cf9ce47d6a.png">

**Comment by Eyesfront-ZJG** [comment]
@p0deje dose the feature develop done?

**Comment by p0deje** [comment]
No, otherwise it would have been closed. On Jul 16, 2023, at 18:39, justdoit ***@***.***> wrote:﻿
@p0deje dose the feature develop done?

**Comment by Eyesfront-ZJG** [comment]
> 

i think the feature is very useful for me。when to release the feature？ thxs!!!

**Comment by gabriel-r** [comment]
IMHO, the specific filter interface is very nice buy overkill.
Simply typing "image" in the search field would be amazing, like in Alfred on my other Mac :)
This could actually apply to other filters, e.g. "file", "text". It could be made smarter with prefixes like "type:image", it's pretty standard, including here on GH.

**Comment by anandghegde** [comment]
Hi @p0deje, I'd like to pick this up. Before writing any code, I want to check which direction you'd merge.

Your segmented-control mockup from 2022 is the fuller version. As a smaller first step, I'm thinking of the search-prefix idea from the last comment:

- `type:image`, `type:file` and `type:text` narrow the list to that kind of item
- the rest of the query still goes through the current search mode (exact/fuzzy/regex/mixed). For example, `type:image invoice` matches images whose OCR text contains "invoice"
- the logic stays in `Search.swift` plus a small `HistoryItem` helper, covered by new cases in `SearchTests`
- no UI changes, so a toggle control could be added later on top of the same filter

To classify items I'd reuse the checks that already exist:

- **file:** `fileURLs` is non-empty
- **image:** `imageData` is present and there are no file URLs
- **text:** everything else

Image-type file URLs from Universal Clipboard (the `universalClipboardImage` case) would count as images.

Two questions:
1. Would you merge the prefix as a first step, or do you only want the toggle UI?
2. When a Finder copy has both a file URL and image data, should it count as a file (my suggestion) or an image?

**Comment by p0deje** [comment]
@anandghegde Yes, I'd merge the prefix as a first step, without any complex UI.

**Comment by anandghegde** [comment]
Thanks @p0deje — opened #1542 with the prefix only, no UI.

`type:image`, `type:file` and `type:text` narrow the list; the rest of the query keeps going through the current search mode, so `type:image invoice` still matches images by their recognized text. Multiple prefixes are a union, and a half-typed `type:fi` is searched for literally rather than emptying the list.

On my second question, which you didn't need to answer: a Finder copy that carries both a file URL and image data is classified as a **file**. One line to flip if you'd rather it counted as an image.
